Add asserts for yielding while holding ticket locks.

SeanTAllen opened the pull request:
Ticket locks are spinlocks, which do not yield.  Any code that yields
while holding a ticket lock can be a source of deadlocks.

SeanTAllen commented:
This is your code David. Even though it hasn't found anything yet, I think it is reasonable to merge these changes as they would help prevent a hard to track down regression in the future.

SeanTAllen commented:
I've never seen this before:

```
2020-06-01T19:20:43.8737882Z -----------stderr-start-------------
2020-06-01T19:20:43.8738616Z [   SGX-LKL  ] SGX-LKL (OE) Git version cc0764d-dirty LKL version 5.3.0 [DEBUG build (-O0)] [SOFTWARE DEBUG]
2020-06-01T19:20:43.8739508Z [   SGX-LKL  ] nproc=8 ETHREADS=8 CMDLINE="mem=32M"
2020-06-01T19:20:43.8740143Z [   SGX-LKL  ] HW TLS support: conf->fsgsbase=0
2020-06-01T19:20:43.8741060Z [   SGX-LKL  ] Registering disk 0 (path='sgx-lkl-rootfs.img', mnt='/', [RW   ])
2020-06-01T19:20:43.8742104Z [   SGX-LKL  ] No tap device specified, networking will not be available.
2020-06-01T19:20:43.8743367Z [   SGX-LKL  ] get_signed_libsgxlkl_path... result=/opt/sgx-lkl-debug/bin/../lib/libsgxlkl.so.signed
2020-06-01T19:20:43.8744489Z [   SGX-LKL  ] oe_create_enclave... result=0 (OE_OK)
2020-06-01T19:20:43.8745457Z [   SGX-LKL  ] sgxlkl_enclave_init(ethread_id=0)
2020-06-01T19:20:43.8746520Z [[  SGX-LKL ]] wg0 has public key 8bfpz0XjrxwxVfIkrGSTAKvGXyN+uZ7VHS1LhSF9Fmk=
2020-06-01T19:20:43.8747199Z free(): invalid pointer
2020-06-01T19:20:43.8747701Z free(): invalid pointer
2020-06-01T19:20:43.8748226Z free(): invalid pointer
2020-06-01T19:20:43.8748701Z free(): invalid pointer
2020-06-01T19:20:43.8749280Z free(): invalid pointer
2020-06-01T19:20:43.8750468Z [   SGX-LKL  ] ethread (4: 0) free(): invalid pointer
2020-06-01T19:20:43.8750867Z Aborted (core dumped)
2020-06-01T19:20:43.8751378Z make: *** [run-sw] Error 134
```

SeanTAllen commented on src/sched/lthread.c:
```diff
@@ -567,8 +567,13 @@ int _lthread_resume(struct lthread* lt)
+        SGXLKL_ASSERT(lt->ticketlocks_held == 0);
     }
 
+    sched->current_lthread = NULL;
```
did you intend to move these two lines?

davidchisnall replied:
Yes, without that any code run in the callback is not able to get the current thread. That makes it impossible for the unlock to decrement the counter.

I did not send this as a PR because I’m uncertain if that has any other negative side effects. I don’t believe it should be possible for the lthread strict to be deallocated before here, but I’m not completely sure. There may be a race condition in the cloned thread deallocation if there are multiple ethreads. That said, the resume function expects the lthread object to be still live because it accesses a field, so this shouldn’t make things worse...

SeanTAllen replied:
Bad things definitely happen with this here. See:

https://dev.azure.com/sgx-lkl/sgx-lkl/_build/results?buildId=1549&view=logs&j=4dfb742b-8e24-5ebe-1c3b-99535daead16

for some segfaults in addition to the assertions being triggered.

SeanTAllen commented:
@davidchisnall closing due to issues. see last note. i don't want to delete the branch as it was yours. Is it ok to delete?

davidchisnall commented:
I recently fixed an issue in the VirtIO code that was sleeping with a ticket lock held. Reopening this to see if we’re closer to being able to enable it.

davidchisnall committed:
Add asserts for yielding while holding ticket locks.

Ticket locks are spinlocks, which do not yield.  Any code that yields
while holding a ticket lock can be a source of deadlocks.

davidchisnall commented:
It looks as if the code in `vio_enclave_wakeup_event_channel` holds a lock across things that can yield (specifically, waking up another thread.  On closer inspection, that code looks as if it can lose wakeups and needs some careful review.
